packages/framework/src/hooks/useStyles.ts

```diff
@@ -0,0 +1,81 @@
+import { useAtomValue } from "jotai";
+import { type CSSProperties, useMemo } from "react";
+import {
+  compact,
+  get,
+  intersection,
+  isEmpty,
+  isString,
+  keys,
+  merge,
+  omit,
+} from "lodash-es";
+import { type EnsembleThemeModel } from "../shared";
+import { themeAtom } from "../state";
+import { useEvaluate } from "./useEvaluate";
+
+const resolveStyleNames = (
+  styleNames: string[],
+  theme: EnsembleThemeModel,
+): CSSProperties => {
+  const appStyleNames = keys(theme.Styles);
+  const overlappingNames = intersection(styleNames, appStyleNames);
+  if (isEmpty(overlappingNames)) {
+    return {};
+  }
+
+  const appliedStyles: CSSProperties = {};
+  overlappingNames.forEach((name) => {
+    merge(appliedStyles, get(theme.Styles, name));
+  });
+
+  return appliedStyles;
+};
+
+export const useStyles = <T extends { [key: string]: unknown }>(
+  values: T,
+): CSSProperties | undefined => {
+  const namedStyles = get(values, ["styles", "names"]) as unknown;
+  const classStyles = get(values, ["styles", "className"]) as unknown;
+
+  const themeContext = useAtomValue(themeAtom);
+
+  const { classStylesArray, nameStylesArray } = useMemo(() => {
+    const stringSplitter = (styles: string): string[] =>
+      // eslint-disable-next-line prefer-named-capture-group
+      styles.split(/([a-zA-Z0-9_-]+)\s+(\${.+?})\s+(\w+)/).filter(Boolean);
+    return {
+      classStylesArray: isString(classStyles)
+        ? stringSplitter(classStyles)
+        : [],
+      nameStylesArray: isString(namedStyles) ? stringSplitter(namedStyles) : [],
+    };
+  }, [classStyles, namedStyles]);
+
+  const {
+    classStylesArray: classStylesArrayEval,
+    nameStylesArray: namedStylesArrayEval,
+  } = useEvaluate({
+    classStylesArray,
+    nameStylesArray,
+  });
+
+  const styleNames = compact([
+    ...namedStylesArrayEval.map((nameStyle) => `${nameStyle}`),
+    ...classStylesArrayEval.map((className) => `.${className}`),
+  ]).join(" ");
```

Comment:
This can be moved inside the next useMemo and there's no point joining and resplitting.

Reply:
No that is wrong, we need to place this outside of useMemo and convert it into string and later resplitting. otherwise this will go into infinite render due to array gives new reference each time.